### PR TITLE
SO-2857: Filter validation issues

### PR DESCRIPTION
**Pull request by norban**

### CODE REVIEW

File: core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationIssueApiTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			.buildAsync()
+			.getRequest()
+			.execute(context);
+		
+		assertThat(issues).hasSize(1);
+		assertThat(issues.stream().map(ValidationIssue::getId).collect(Collectors.toList())).containsOnly(issueWithDetail);
+		
+	}
+
+	private Map<String, Object> createDetails(String key, Object value) {
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationIssueApiTest.java]: Replace this method with `ImmutableMap.<String, Object>of(key, value)`.

File: core/com.b2international.snowowl.datastore/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -27,6 +27,7 @@ Require-Bundle: org.eclipse.core.runtime;bundle-version="3.9.0",
  com.b2international.scripting.api;visibility:=reexport
 Export-Package: com.b2international.snowowl.core.internal.validation,
  com.b2international.snowowl.core.validation,
+ com.b2international.snowowl.core.validation.detail,
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/META-INF/MANIFEST.MF]: Merge the contents of the detail package into the issue package.

File: core/com.b2international.snowowl.datastore/plugin.xml [hunk 3]
```diff
@@ -9,6 +9,7 @@
    <extension-point id="versioningManager" name="versioningManager" schema="schema/versioningManager.exsd"/>
    <extension-point id="contentAvailabilityInfoProvider" name="Content Availability Info Provider" schema="schema/contentAvailabilityInfoProvider.exsd"/>
    <extension-point id="mappingProvider" name="mappingProvider" schema="schema/mappingProvider.exsd"/>
+   <extension-point id="validationDetailExtension" name="validationDetailExtension" schema="schema/validationDetailExtension.exsd"/>
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/plugin.xml]: Rename extension point to `validationIssueDetailExtension`.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java [hunk 4]
```diff
@@ -39,7 +40,7 @@
 /**
  * @since 6.0
  */
-final class ValidateRequest implements Request<BranchContext, ValidationResult> {
+final class ValidateRequest implements Request<BranchContext, ValidationResult>{
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/ValidateRequest.java]: Revert this change.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/IssueDetail.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+package com.b2international.snowowl.core.validation.detail;
+
+import java.util.Map;
+
+import com.b2international.snowowl.core.ComponentIdentifier;
+
+/**
+ * @since 6.4
+ */
+public class IssueDetail {
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/IssueDetail.java]: Make this class `final`.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtension.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package com.b2international.snowowl.core.validation.detail;
+
+import com.b2international.commons.options.Options;
+import com.b2international.index.query.Expressions.ExpressionBuilder;
+
+/**
+ * @since 6.4
+ */
+public interface ValidationDetailExtension {
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtension.java]: Rename to `ValidationIssueDetailExtension`.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtensionProvider.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	
+	public Collection<ValidationDetailExtension> getExtensions() {
+		return extensions;
+	}
+	
+	/**
+	 * Usage of this function is intended for testing purposes only
+	 */
+	public void addExtension(ValidationDetailExtension extension) {
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtensionProvider.java]: If possible restrict the visibility of this method to `package` private.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java [hunk 8]
```diff
@@ -52,6 +58,8 @@
 	private final short affectedComponentType;
 	private final boolean whitelisted;
 	
+	private Map<String, Object> details = newHashMap();
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java]: Use `Collections.emptyMap()` or `null` here so fetching issues without details does not create a bunch of empty maps.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java [hunk 9]
```diff
@@ -113,13 +121,28 @@ public boolean isWhitelisted() {
...
 	}
 	
+	@JsonAnyGetter
+	public Map<String, Object> getDetails() {
+		return details;
+	}
+	
+	@JsonAnySetter
+	public void setDetails(String key, Object value) {
+		this.details.put(key, value);
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java]: Make sure you have a mutable details field here.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java [hunk 10]
```diff
@@ -113,13 +121,28 @@ public boolean isWhitelisted() {
...
+	}
+	
+	public void setDetails(Map<String, Object> details) {
+		this.details.putAll(details);
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssue.java]: And here as well.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssueSearchRequestBuilder.java [hunk 11]
```diff
@@ -74,9 +77,14 @@ public ValidationIssueSearchRequestBuilder isWhitelisted(boolean whitelisted) {
 		return addOption(OptionKey.WHITELISTED, whitelisted);
 	}
 	
+	public ValidationIssueSearchRequestBuilder filterByDetails(Map<String, Object> details) {
+		final Options options = Options.builder().putAll(details).build();
```
Comment (cmark) [inline core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/issue/ValidationIssueSearchRequestBuilder.java]: Add a static `from(Map): Options` method to `Options`.

File: snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/validation/SnomedValidationIssueDetailTest.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+				.all()
+				.filterByDetails(details)
+				.buildAsync()
+				.getRequest()
+				.execute(context);
+		
+		assertComponents(issues, issueWithActiveComponent.getAffectedComponent());
+	}
+	
+	private Map<String, Object> createDetails(String key, Object value) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/validation/SnomedValidationIssueDetailTest.java]: Replace with `ImmutableMap.<String, Object>of(key, value)`.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java [hunk 13]
```diff
@@ -63,17 +67,24 @@
...
 				.build()
 				.execute(context)
 				.stream()
-				.map(SnomedComponent::getComponentIdentifier)
+				.map(component -> {
+					final ComponentIdentifier affectedComponent = component.getComponentIdentifier();
+					final String moduleId = component.getModuleId();
+					final Map<String, Object> details = newHashMap();
+					details.put(SnomedRf2Headers.FIELD_MODULE_ID, moduleId);
+					details.put(SnomedRf2Headers.FIELD_ACTIVE, Boolean.valueOf(component.isActive()));
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: `component.isActive()` already returns a wrapper object, no need for `Boolean.valueOf()` call.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/detail/SnomedValidationDetailExtension.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+	@Override
+	public void prepareQuery(ExpressionBuilder queryBuilder, Options options) {
+		if (options.containsKey(SnomedRf2Headers.FIELD_ACTIVE)) {
+			final Boolean isActive = options.get(SnomedRf2Headers.FIELD_ACTIVE, Boolean.class);
+			queryBuilder.filter(Expressions.match(SnomedRf2Headers.FIELD_ACTIVE, isActive));
+		}
+		
+		if (options.containsKey(SnomedRf2Headers.FIELD_MODULE_ID)) {
+			final Collection<String> moduleIds = options.getCollection(SnomedRf2Headers.FIELD_MODULE_ID, String.class);
+			if (moduleIds.isEmpty()) {
```
Comment (cmark) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/detail/SnomedValidationDetailExtension.java]: Why do you check `isEmpty()` here?
Searching for empty values should return empty results.

**Review by cmark — CHANGES_REQUESTED** [review]

**Comment by cmark** [comment]
Also make sure you migrate the new validation rules added since you've started this branch. Check the `com.b2international.snowowl.validation.snomed` project in this repository for example.

File: core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtensionProvider.java (hunk 7 above)
Comment (norban) [reply core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/detail/ValidationDetailExtensionProvider.java]: Couldn't do it since we need it in the test classes

**Comment by norban** [comment]
Requested changes have been applied

**Review by cmark — APPROVED** [review]
LGTM!

**Review by cmark — APPROVED** [review]
LGTM!
